src/ports/postgres/modules/graph/hits.py_in

```diff
@@ -0,0 +1,417 @@
+            # calculate authority
+            # if there is no node that point to A, authority_i(A) = 0
+            ###################################################################
+            plpy.execute("""
+                    CREATE TABLE {message} AS
+                    SELECT {cur}.{vertex_id} AS {vertex_id},
+                            COALESCE(SUM({v1}.hub), 0.0) AS authority,
+                            {cur}.hub AS hub
+                    FROM {cur}
+                        LEFT JOIN {edge_temp_table} ON {cur_join_clause}
+                        LEFT JOIN {cur} AS {v1} ON {v1_join_clause}
+                    GROUP BY {cur}.{vertex_id}, {cur}.hub
+                    ORDER BY {cur}.{vertex_id}
+                    {cur_distribution}
+                """.format(**locals()))
+
+            #####################################################
+            # calculate hub
+            # if node A doesn't point to any node, hub_i(A) = 0
+            #####################################################
+            message_join_clause = "{message}.{vertex_id} = {edge_temp_table}.{src}".format(
+                **locals())
+            v2_join_clause = "{v2}.{vertex_id} = {edge_temp_table}.{dest}".format(
+                **locals())
+            plpy.execute("""
+                    UPDATE {message}
+                    SET hub = {tmp2}.hub
+                    FROM
+                    (SELECT {message}.{vertex_id} AS {vertex_id},
+                            COALESCE(SUM({v2}.authority), 0) AS hub
+                    FROM {message}
+                        LEFT JOIN {edge_temp_table} ON {message_join_clause}
+                        LEFT JOIN {message} AS {v2} ON {v2_join_clause}
+                    GROUP BY {message}.{vertex_id}) AS {tmp2}
+                    WHERE {tmp2}.{vertex_id} = {message}.{vertex_id}
+                """.format(**locals()))
+
+            # normalize authority value with L2 distance
+            authority_norm = math.sqrt(plpy.execute("""
+                    SELECT SUM({tmp}.authority_square) AS sum_norm_square
+                    FROM
+                    (SELECT POWER(authority, 2) AS authority_square
+                    FROM {message}) AS {tmp}
+                """.format(**locals()))[0]["sum_norm_square"])
+
+            if authority_norm != 0:
+                plpy.execute("""
+                        UPDATE {message}
+                        SET authority = authority/{authority_norm}
+                    """.format(**locals()))
+
+            # normalize hub value with L2 distance
+            hub_norm = math.sqrt(plpy.execute("""
+                    SELECT SUM({tmp3}.hub_square) AS sum_hub_square
+                    FROM
+                    (SELECT POWER(hub, 2) AS hub_square
+                    FROM {message}) AS {tmp3}
+                """.format(**locals()))[0]["sum_hub_square"])
+
+            if hub_norm != 0:
```

Comment:
`hub_norm == 0` and `authority_norm == 0` are error cases. You could probably check for `if hub_norm == 0 or authority_norm == 0:` and exit the program with an error message on True.

Reply:
Yes, this makes more sense.